notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,452 @@
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We notice a stark contrast in the predictions with males being a lot more likely to be predicted to get approved, similar to the original training data. However, there's even a disparity between the subgroup of males and females that are approved. When considering only the samples labeled with \"loan not repaid\" males are much more likely to be predicted to get approved than females. The test data shows a similar disparity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.widget import FairlearnDashboard\n",
+    "\n",
+    "FairlearnDashboard(sensitive_features=sensitive_features_test.values,\n",
+    "                   true_y=y_test.values,\n",
+    "                   predicted_ys=unconstrained_predictor.predict(X_test),\n",
+    "                   class_names=None,\n",
+    "                   feature_names=X_test.columns.values.tolist(),\n",
+    "                   is_classifier=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Postprocessing the model to get a fair model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The idea behind postprocessing is to alter the output of the fairness-unaware model to achieve fairness. The postprocessing algorithm requires three input arguments:\n",
+    "- the matrix of samples X\n",
+    "- the vector of predictions y from the fairness-unaware model \n",
+    "- the vector of group attribute values A (in our code we refer to it as `sensitive_features`)\n",
+    "\n",
+    "The goal is to make the output fair with respect to constraints. Our postprocessing algorithm uses one of\n",
+    "- Demographic Parity (DP): $P[h(X)=\\hat{y} | A=a] = P[h(X)=\\hat{y}] \\qquad \\forall a, \\hat{y}$\n",
+    "- Equalized Odds (EO): $P[h(X)=\\hat{y} | A=a, Y=y] = P[h(X)=\\hat{y}|Y=y] \\qquad \\forall a, \\hat{y}$\n",
+    "\n",
+    "where $h(X)$ is the prediction based on the input $X$, $\\hat{y}$ and $y$ are labels, and $a$ is a sensitive feature value. In our example, we'd expect the postprocessed model with DP to be balanced between races. In this particular scenario it makes more sense to aim at fairness through accuracy like EO. EO does not make the same guarantees. Instead, it ensures that the parity between the subgroups of each race with label 1 in the training set, and parity between the subgroups of each race with label 0 in the training set. Applied to our scenario, this means that the subgroup of each race who reoffended in the past are equally likely to be predicted to reoffend (and therefore also equally likely not to). Similarly, there is parity between subgroups of each race without recidivism, but we have no parity between the groups with different training labels. In mathematical terms at the example of African-American and Caucasian:\n",
```

Comment:
I think the sentence makes more sense without 'that the'

Reply:
"this means subgroup of each race" doesn't really work, though.

I can make it "this means the reoffending subgroups of each race are equally....."

Reply:
oh there were multiple "that the"s, and you likely meant the first one... done!